sklearn/neighbors/quad_tree.pyx

```diff
@@ -0,0 +1,547 @@
+        return res
+
+
+    cdef SIZE_t select_child(self, DTYPE_t[3] point, Cell* cell) nogil:
+        cdef int i
+        cdef SIZE_t selected_child = 0
+        for i in range(self.n_dimensions):
+            # Select the correct child cell to insert the point by comparing
+            # it to the borders of the cells using precomputed center.
+            selected_child *= 2
+            if point[i] >= cell.center[i]:
+                selected_child += 1
+        return cell.children[selected_child]
+
+    cdef void _init_root(self, DTYPE_t[3] min_bounds, DTYPE_t[3] max_bounds) nogil:
+        cdef int i
+        cdef DTYPE_t width
+        cdef Cell* root = &self.cells[0]
+        self.init_cell(root, -1, 0)
+        for i in range(self.n_dimensions):
+            root.min_bounds[i] = min_bounds[i]
+            root.max_bounds[i] = max_bounds[i]
+            root.center[i] = (max_bounds[i] + min_bounds[i]) / 2.
+            width = max_bounds[i] - min_bounds[i]
+            root.max_width = max(root.max_width, width*width)
+        root.cell_id = 0
+        
+        self.cell_count += 1
+
+    def build_tree(self, X):
+        """Build a tree from the points in X."""
+        cdef DTYPE_t[3] pt
+        cdef DTYPE_t[3] min_bounds, max_bounds
+
+        # validate X and prepare for query
+        # X = check_array(X, dtype=DTYPE_t, order='C')
+        n_samples = X.shape[0]
+
+        capacity = 100
+        self._resize(capacity)
+        m, M = np.min(X, axis=0) - 1e-3, np.max(X, axis=0) + 1e-3
+        for i in range(self.n_dimensions):
+            min_bounds[i] = m[i]
+            max_bounds[i] = M[i]
+
+        # Create the initial node with boundaries from the dataset
+        self._init_root(min_bounds, max_bounds)
+
+        for i in range(n_samples):
+            for j in range(self.n_dimensions):
+                pt[j] = X[i, j]
+            self.insert_point(pt, i)
+
+        self._resize(capacity=self.cell_count)
+
+    def plot_tree(self):
+        """Plot the tree with cell boundaries and the points inserted in it."""
+        self.check_coherence()
+        import matplotlib.pyplot as plt
+
+        plt.figure()
+        for c in self.cells[:self.cell_count]:
+            if not c.is_leaf:
+                # Plot the cell division if the cell is an inner cell
+                plt.vlines(c.center[0], c.min_bounds[1], c.max_bounds[1])
+                plt.hlines(c.center[1], c.min_bounds[0], c.max_bounds[0])
+            else:
+                # If the cell is a leaf, display the point contained in it.
+                plt.scatter(c.barycenter[0], c.barycenter[1], c='b', marker='.')
+            
+        # Print bounding box of the Tree
+        root = self.cells[0]
+        plt.vlines([root.min_bounds[0], root.max_bounds[0]], root.min_bounds[1], root.max_bounds[1])
+        plt.hlines([root.min_bounds[1], root.max_bounds[1]], root.min_bounds[0], root.max_bounds[0])
+        plt.show()
```

Comment:
For people interested, here is the result of running this on a `make_blobs` dataset with 1000 points and 5 clusters:

![image](https://user-images.githubusercontent.com/89061/27493345-d660e2ec-5849-11e7-8d8b-3da1aaa043b4.png)

It seems to work as expected.